---
title: "Homework Assignment"
subtitle: "Pair-wise Sequence Alignment"
output:
  pdf_document: default
date: "03-04-2023"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE)
knitr::opts_chunk$set(cache=TRUE)
knitr::opts_chunk$set(tidy=TRUE)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff = 60))
source("~/bioinformatics/hw3/hw3_func.R")

require(seqinr)
require(Biostrings)
require(ORFhunteR)
require(gridExtra)
require(ggplot2)

choosebank("genbank")
```

# Task A
```{r taskA}
# Accession numbers
ac_1 <- "AY884001"
ac_2 <- "MH940245"

# Retrieve the data
q1 <- query("q1", paste("AC=", ac_1))
q2 <- query("q2", paste("AC=", ac_2))

# Get the sequences
seq1 <- getSequence(q1$req[[1]])
seq2 <- getSequence(q2$req[[1]])

# Print the sequences
print(DNAString(c2s(seq1)))
print(DNAString(c2s(seq2)))
```

# Task B
```{r taskB}
p1 <- seqinr::translate(seq1)
p2 <- seqinr::translate(seq2)

seq1_count <- table(p1)
seq2_count <- table(p2)
print(seq1_count)
print(seq2_count)

seq1_prop <- proportions(seq1_count)
seq2_prop <- proportions(seq2_count)
print(seq1_prop)
print(seq2_prop)

theme_update(legend.key.size = unit(3, "mm"), legend.text = element_text(size = 6), axis.text = element_blank(), axis.ticks = element_blank(), panel.grid  = element_blank())
df1 <- data.frame(slices = seq1_prop, amino_acids = paste(names(seq1_prop), "=", round(seq1_prop * 100, 2), "%", sep = ""))
df2 <- data.frame(slices = seq2_prop, amino_acids = paste(names(seq2_prop), "=", round(seq2_prop * 100, 2), "%", sep = ""))
plot1 <- ggplot(df1, aes(x = "", y = slices.Freq, fill = amino_acids)) + geom_bar(stat = "identity", width = 1) + coord_polar("y", start = 0) + ggtitle(ac_1) + xlab("") + ylab("") + guides(fill = guide_legend(title = NULL))
plot2 <- ggplot(df2, aes(x = "", y = slices.Freq, fill = amino_acids)) + geom_bar(stat = "identity", width = 1) + coord_polar("y", start = 0) + ggtitle(ac_2) + xlab("") + ylab("") + guides(fill = guide_legend(title = NULL))
grid.arrange(plot1, plot2, ncol = 2)
```


# Task C
```{r taskC}
seq1_orfs <- findORFs(c2s(seq1))
seq2_orfs <- findORFs(c2s(seq2))

dotPlot(s2c(seq1_orfs[1, "orf.sequence"]), s2c(seq2_orfs[1, "orf.sequence"]), xlab = ac_1, ylab = ac_2)
```

The symmetric nature of the plot reveals that the two sequences are almost identical. In fact, by doing a direct comparison
we can see that they are exactly the same!
```{r taskC1}
print(seq1_orfs[1, "orf.sequence"] == seq2_orfs[1, "orf.sequence"])
```


# Task D
```{r taskD}
opt_ga <- pairwiseAlignment(DNAString(c2s(seq1)), DNAString(c2s(seq2)), type = "global", substitutionMatrix = nucleotideSubstitutionMatrix(match = 2, mismatch = -1, baseOnly = TRUE), gapOpening = 0, gapExtension = -2)
print(opt_ga)

cat("Alignment for first 20 nucleotides:", paste(substr(pattern(opt_ga), 1, 20), substr(subject(opt_ga), 1, 20), sep = '\n'), sep = "\n")
```

# Task E
```{r taskE}
seq1_random <- generateSeqsWithMultinomialModel(c2s(seq1), 20)
seq2_random <- generateSeqsWithMultinomialModel(c2s(seq2), 20)

scores <- c()
for (i in 1:20) {
  score <- pairwiseAlignment(DNAString(c2s(seq1_random[[i]])), DNAString(c2s(seq2_random[[i]])), type = "global", substitutionMatrix = 
nucleotideSubstitutionMatrix(match = 2, mismatch = -1, baseOnly = TRUE), gapOpening = 0, gapExtension = -2, scoreOnly = TRUE)
  scores <- append(scores, score)
}

print(scores)
hist(scores)

p_value <- sum(scores > score(opt_ga)) / length(scores)
print(p_value)
```
The global alignment is statistically significant as the p-value is below 0.05. This means that the odds of this alignment 
happening by chance are extremely low.

# Task F
```{r taskF}
opt_la <- pairwiseAlignment(DNAString(c2s(seq1)), DNAString(c2s(seq2)), type = "local", substitutionMatrix = nucleotideSubstitutionMatrix(match = 3, mismatch = -2, baseOnly = TRUE), gapOpening = -4, gapExtension = -2)
print(opt_la)
print("Length of the alignment:")
print(nchar(pattern(opt_la)))
```
